File: sdks/python/kanon_sdk/context.py
"""Context and message segment abstractions for Kanon Python SDK."""

import asyncio
import uuid
from pathlib import Path
from typing import Any, Dict, Optional

from google.protobuf.json_format import ParseDict
from google.protobuf.struct_pb2 import Struct

from kanon_sdk.proto import pb, pb_grpc


class CoreHandle:
    """Adapter-facing handle for pushing inbound events into the Kanon Core pipeline.

    A handle wraps one *already connected* ``BotApiServiceStub``; it never dials,
    never re-dials and never closes a channel itself. Connection ownership stays
    with the plugin host, which keeps exactly one IPC endpoint and one lifecycle
    in the process.

    Why a single shared channel instead of dialing per call
    -------------------------------------------------------
    Adapters ingest at message rate, and every ``grpc.aio.insecure_channel``
    allocates a fresh HTTP/2 connection (file descriptor, connectivity task and
    handshake round trips) to ``core.sock``. Dialing per ``ingest_event`` call
    would multiply descriptors and latency for no benefit, and would leave
    orphaned channels behind whenever a task is cancelled mid-flight. Core
    exposes one endpoint and multiplexes concurrent streams over HTTP/2, so a
    single long-lived channel is both the cheapest and the intended topology;
    it also reconnects on its own after a Core restart, so adapters need no
    dialing or reconnection logic of their own.

    Because several adapter tasks (one poller per platform/chat) may share that
    channel, the handle serializes ingest calls behind an ``asyncio.Lock``:

    * ``accepted == False`` backpressure stays actionable - one call in flight at
      a time means a "queue full" answer is never masked by a concurrent burst;
    * events reach Core in the same order the adapter tasks scheduled them,
      which matters for per-channel conversational ordering;
    * the shared stub is never used concurrently with host shutdown, so closing
      the channel cannot race an in-flight RPC from another task.

    Fast-ACK and why ``IngestEventResponse.accepted`` matters
    ---------------------------------------------------------
    Core never blocks on the pipeline: ``IngestEvent`` performs a non-blocking
    enqueue into a bounded queue and answers immediately. A response therefore
    means *"Core received your event"*, not *"Core processed your event"*. When
    the queue is full Core returns ``accepted=False`` and drops the event to
    protect the microkernel; adapters must inspect that flag (and may retry later
    at their own discretion). This handle deliberately returns the raw response
    instead of raising or retrying, so the caller owns that policy.

    Transport failures are *not* converted into a response: if Core is
    unreachable, the RPC raises ``grpc.aio.AioRpcError``. The SDK never
    fabricates ``accepted=True`` (or any synthetic response) for a call that did
    not reach Core.

    Usage (platform adapter plugin)::

        import asyncio

        import grpc

        from kanon_sdk import Plugin, PluginContext


        class TelegramAdapter(Plugin):
            async def on_load(self, ctx: PluginContext) -> None:
                # Capture the handle once: ctx.core is None in standalone mode,
                # i.e. when the host could not reach Core at startup.
                self.core = ctx.core
                self._poller = asyncio.create_task(self._poll_updates())

            async def _poll_updates(self) -> None:
                while True:
                    update = await telegram_get_updates()
                    if self.core is None:
                        continue  # Standalone mode: nothing to ingest into.
                    try:
                        resp = await self.core.ingest_event(
                            platform="telegram",
                            channel_id=str(update.chat_id),
                            sender_id=str(update.user_id),
                            text=update.text,
                            metadata={"update_id": update.update_id},
                        )
                    except grpc.aio.AioRpcError:
                        continue  # Core was not reached; retry on the next poll.
                    if not resp.accepted:
                        continue  # Fast-ACK backpressure: Core queue is full.

            async def on_deliver_message(self, req):
                # Adapters must also override the outbound hook; see Plugin.
                ...
    """

    def __init__(self, stub: pb_grpc.BotApiServiceStub) -> None:
        """Wraps an existing ``BotApiService`` stub without touching its channel."""
        # The stub (and therefore the channel) belongs to the host process: the
        # handle only issues RPCs, so channel creation/closure has a single owner.
        self._stub = stub
        # Serializes concurrent adapter tasks sharing this handle. The lock is
        # created lazily against the running loop, so constructing a CoreHandle
        # outside a loop (e.g. at import time) stays safe.
        self._lock = asyncio.Lock()

    async def ingest_event(
        self,
        platform: str,
        channel_id: str,
        sender_id: str,
        text: str,
        event_id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> pb.IngestEventResponse:
        """Pushes one inbound platform message into the Core pipeline.

        Args:
            platform: Adapter platform identifier, e.g. ``"telegram"``.
            channel_id: Platform-side conversation/channel/group identifier.
            sender_id: Platform-side identifier of the message author.
            text: Raw inbound text. Rich media is intentionally out of scope for
                this primitive; adapters may describe it through ``metadata``.
            event_id: Optional caller-supplied idempotency/deduplication id. When
                omitted a random ``uuid4().hex`` is generated, so every call
                still carries a non-empty id.
            metadata: Optional JSON-compatible mapping forwarded to Core as a
                ``google.protobuf.Struct`` (nested dicts/lists are supported).

        Returns:
            The raw ``pb.IngestEventResponse``. ``accepted`` reflects Core's
            Fast-ACK decision: ``False`` means the event was *not* enqueued
            because the ingest queue was full, so callers must not assume
            success. ``event_id`` echoes the id used for this event.

        Raises:
            grpc.aio.AioRpcError: If the shared channel could not reach Core, or
                Core answered with an error status. Deliberately not swallowed:
                "Core unreachable" and "Core rejected the event" are different
                outcomes and the adapter decides how to react.
        """
        # Always ship a non-empty event id: Core logs and returns it verbatim, and
        # an empty id would make deduplication and tracing impossible.
        resolved_event_id = event_id or uuid.uuid4().hex

        # Convert the caller's mapping into a Struct up front, using the same
        # JSON mapping helper as Plugin.meta()/on_call_tool so payload conversion
        # follows one rule across the SDK. ParseDict fails loudly on values that
        # JSON cannot express, which is preferable to silently dropping fields.
        event_metadata: Optional[Struct] = None
        if metadata is not None:
            event_metadata = Struct()
            ParseDict(metadata, event_metadata)

        # Both the outer request and the nested event carry the platform: Core's
        # pipeline worker prefers event.platform and only falls back to the outer
        # field, so setting both keeps routing unambiguous.
        request = pb.IngestEventRequest(
            platform=platform,
            event=pb.PipelineEventRequest(
                event_id=resolved_event_id,
                platform=platform,
                channel_id=channel_id,
                sender_id=sender_id,
                raw_text=text,
                metadata=event_metadata,
            ),
        )

        # Single-flight per handle: see the class docstring for why adapter tasks
        # share one channel and one lock instead of dialing concurrently. The
        # await is short by design - Core Fast-ACKs without waiting for the LLM.
        async with self._lock:
            return await self._stub.IngestEvent(request)


class PluginContext:
    """Runtime context provided to a plugin during initialization and execution.

    Attributes:
        data_dir: Writable per-plugin data directory (``./data/plugins/<id>/``).
        config: Static plugin configuration mapping (empty when absent).
        core: Optional :class:`CoreHandle` for pushing inbound events into Core.
            Built by the host from the same channel used for ``RegisterHost``.
            It is ``None`` in standalone mode - when ``KANON_CORE_SOCK`` is
            unset, its socket is missing, or registration failed - and plugins
            must treat that as "no Core available" rather than dialing their own.
    """

    def __init__(
        self,
        data_dir: Path,
        config: Optional[Dict[str, Any]] = None,
        core: Optional[CoreHandle] = None,
    ):
        """Builds a context; ``config`` and ``core`` are optional keywords."""
        self.data_dir = data_dir
        self.config = config or {}
        # Never a fabricated stub: either the host proved Core reachable, or None.
        self.core = core


class MessageSegment:
    """Helper factory for constructing strongly typed Protobuf MessageSegment instances."""

    @staticmethod
    def text(content: str) -> pb.MessageSegment:
        """Constructs a plain text segment."""
        return pb.MessageSegment(text=pb.TextSegment(content=content))

    @staticmethod
    def image_url(
        url: str,
        mime_type: Optional[str] = None,
        filename: Optional[str] = None,
    ) -> pb.MessageSegment:
        """Constructs an image segment referencing a remote URL."""
        return pb.MessageSegment(
            image=pb.ImageSegment(
                url=url,
                mime_type=mime_type,
                filename=filename,
            )
        )

    @staticmethod
    def image_file(
        file_path: str,
        mime_type: Optional[str] = None,
        filename: Optional[str] = None,
    ) -> pb.MessageSegment:
        """Constructs an image segment referencing a local zero-copy physical file."""
        return pb.MessageSegment(
            image=pb.ImageSegment(
                file_path=file_path,
                mime_type=mime_type,
                filename=filename,
            )
        )

    @staticmethod
    def image_bytes(
        raw_bytes: bytes,
        mime_type: Optional[str] = None,
        filename: Optional[str] = None,
    ) -> pb.MessageSegment:
        """Constructs an image segment from raw byte payload."""
        return pb.MessageSegment(
            image=pb.ImageSegment(
                raw_bytes=raw_bytes,
                mime_type=mime_type,
                filename=filename,
            )
        )

File: sdks/rust/kanon-sdk/src/context.rs
//! Plugin execution context, environment metadata, and the core IPC handle.
//!
//! Exposes access to isolated persistent directories, active plugin configurations, and — for
//! plugins acting as platform adapters — a [`CoreHandle`] that pushes inbound messages back into
//! the microkernel pipeline.

use std::path::PathBuf;
use std::sync::Arc;

use kanon_proto::v1::bot_api_service_client::BotApiServiceClient;
use kanon_proto::v1::{
    IngestEventRequest, IngestEventResponse, PipelineEventRequest, RegisterHostRequest,
    RegisterHostResponse,
};
use tokio::sync::Mutex;
use tonic::transport::Channel;

/// Runtime context passed to plugins during initialization and invocation.
#[derive(Debug, Clone)]
pub struct PluginContext {
    /// Dedicated directory for this plugin's local database and file storage (`./data/plugins/<id>/`).
    pub data_dir: PathBuf,
    /// Active dynamic configuration parameters provided by Core.
    pub config: Option<prost_types::Struct>,
    /// Handle to the Core microkernel, absent when the host runs standalone.
    ///
    /// Adapter plugins must capture this handle during `on_load` (the only hook receiving the
    /// context) and reuse it from their polling loops:
    /// ```ignore
    /// async fn on_load(&mut self, ctx: &mut PluginContext) -> PluginResult<()> {
    ///     self.core = ctx.core.clone();
    ///     Ok(())
    /// }
    /// ```
    pub core: Option<CoreHandle>,
}

impl PluginContext {
    /// Creates a new `PluginContext` for the given data directory.
    pub fn new(data_dir: PathBuf, config: Option<prost_types::Struct>) -> Self {
        Self {
            data_dir,
            config,
            core: None,
        }
    }

    /// Attaches a Core IPC handle, enabling inbound ingestion from this plugin.
    pub fn with_core(mut self, core: CoreHandle) -> Self {
        self.core = Some(core);
        self
    }
}

/// Cloneable handle for calling back into the Core microkernel (`BotApiService`).
///
/// A single gRPC channel is shared across every clone instead of dialing per call: plugin hosts
/// issue these calls from long-running loops (a chat platform poller, a socket reader), and
/// re-establishing a connection per message would dominate the cost of a text-only event.
#[derive(Debug, Clone)]
pub struct CoreHandle {
    client: Arc<Mutex<BotApiServiceClient<Channel>>>,
}

/// `tonic::Status` is inherently large (it carries metadata and a boxed source), and boxing it
/// here would force every plugin author to unwrap a second layer for no benefit. The rest of the
/// workspace follows the same convention.
#[allow(clippy::result_large_err)]
impl CoreHandle {
    /// Wraps an established core channel.
    pub fn new(channel: Channel) -> Self {
        Self {
            client: Arc::new(Mutex::new(BotApiServiceClient::new(channel))),
        }
    }

    /// Pushes an inbound event into the core pipeline and returns the Fast-ACK acknowledgement.
    ///
    /// The returned [`IngestEventResponse`] carries `accepted`: when the core's ingest queue is
    /// saturated it answers `accepted = false` instead of blocking, and the adapter decides
    /// whether to retry or drop. Transport failures surface as `tonic::Status` errors — this
    /// method never reports success for a message the core did not accept.
    pub async fn ingest_event(
        &self,
        request: IngestEventRequest,
    ) -> Result<IngestEventResponse, tonic::Status> {
        let mut client = self.client.lock().await;
        let response = client.ingest_event(request).await?;
        Ok(response.into_inner())
    }

    /// Answers a liveness probe from the core.
    ///
    /// Deliberately trivial: it opens no business path, so a host can distinguish "core is gone"
    /// from "core is busy". Used by [`crate::watchdog::watch_core`].
    pub async fn ping(&self) -> Result<(), tonic::Status> {
        let mut client = self.client.lock().await;
        let request = kanon_proto::v1::PingRequest {
            timestamp: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .map(|d| d.as_millis() as i64)
                .unwrap_or_default(),
        };
        client.ping(request).await?;
        Ok(())
    }

    /// Registers the hosting process with the core (`BotApiService.RegisterHost`).
    ///
    /// Used by the host runner during startup; exposed because registration is also the cheapest
    /// proof that the core endpoint is actually serving.
    pub async fn register_host(
        &self,
        request: RegisterHostRequest,
    ) -> Result<RegisterHostResponse, tonic::Status> {
        let mut client = self.client.lock().await;
        let response = client.register_host(request).await?;
        Ok(response.into_inner())
    }

    /// Convenience wrapper building a text-only event for `platform`.
    ///
    /// `event_id` is supplied by the caller because platform message identifiers (update ids,
    /// message ids) are the most useful trace keys; pass an empty string to let the core accept
    /// the event without a caller-assigned identifier.
    pub async fn ingest_text(
        &self,
        platform: impl Into<String>,
        channel_id: impl Into<String>,
        sender_id: impl Into<String>,
        text: impl Into<String>,
    ) -> Result<IngestEventResponse, tonic::Status> {
        let platform = platform.into();
        let request = IngestEventRequest {
            platform: platform.clone(),
            event: Some(PipelineEventRequest {
                event_id: String::new(),
                platform,
                channel_id: channel_id.into(),
                sender_id: sender_id.into(),
                raw_text: text.into(),
                segments: Vec::new(),
                metadata: None,
            }),
        };

        self.ingest_event(request).await
    }
}
